### ISSUE TITLE
FR: support inline documentation for classes

### ISSUE THREAD

**Issue by aspiers** [supercollider/supercollider#3630]
## Description

Currently documentation of any class lives in a separate `.schelp` file to the class itself.  I propose that `sclang` (and `scide` if necessary) be enhanced to also allow documentation embedded within classes.  Of course this should not prevent the documentation from being rendered in the same way it currently is, i.e. to HTML and to `scide`'s help browser.

I know this is a big ask, so I merely ask that it be *considered* as a possible direction to take in the long-term.  I'm sure it would be a lot of work, so I'm not expecting it to happen any time soon ;-)

## Rationale

- Tighter coupling of documentation to code encourages better documentation:

  When modifying a method, if the documentation is immediately above the code, it's a lot harder to ignore the docs and only modify the code.  Therefore the risk of forgetting to update documentation to match code changes is much lower.  This applies equally to the case where the developer thinks "oh, I'll update the docs later in a separate commit" and then never gets around to it.

  Similarly, when adding a method to a class which is already documented, it's a lot harder to feel OK about skipping the documentation if it would be polluting an otherwise fully documented class with an undocumented method.  And of course, when deleting a method, it's a no brainer to delete the documentation which is right next to it, preventing accidental creation of orphaned documentation.

- Inline documentation makes code easier to read

  This one should be self-explanatory.  It's easier to understand code when the purpose of the method and the details of the arguments etc. are provided immediately above it.

- Emulates best practice familiar to developers from other languages

  Inline documentation is of course standard practice in many other languages.  This is why projects like [Doxygen](http://www.stack.nl/~dimitri/doxygen/) are so popular.

## Out of scope

Non-class documentation, such as guides, tutorials etc. would not need to be touched - they could still live in separate `.schelp` files as they do now.

**Comment by telephon** [comment]
Recently, there was a discussion on sc-users about this, and we don't agree that inline documentation is a good thing. I personally prefer to read source code, and find that it makes the source code harder to read.

**Comment by aspiers** [comment]
[**@telephon**](https://github.com/telephon) commented on [4 Apr 2018, 12:02 BST](https://github.com/supercollider/supercollider/issues/3630#issuecomment-378562148 "2018-04-04T11:02:10Z - Replied by Github Reply Comments"):
> Recently, there was a discussion on sc-users about this

Any chance of a link?  I can't find that.

> and we don't agree that inline documentation is a good thing. I personally prefer to read source code, and find that it makes the source code harder to read.

I'm always open to alternative view points, but please could you state the rationale behind your preference, like I did above?  Otherwise it's impossible to objectively weigh the pros and cons, and it's not very productive to debate subjective preferences.

**Comment by brianlheim** [comment]
> we don't agree that inline documentation is a good thing

I definitely do not feel this way, and I think giving users the option to document in-source would be worthwhile.

**Comment by brianlheim** [comment]
> > Recently, there was a discussion on sc-users about this

> Any chance of a link? I can't find that.

I think he is talking about this thread:

http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/Doxygen-comments-in-C-source-td7637837.html

See also the discussion here: https://github.com/supercollider/supercollider/issues/2897

**Comment by aspiers** [comment]
Thanks :-)

**Comment by telephon** [comment]
Btw. I think that inline documentation is less disputed in the c++ and c parts. The main point (at least to me) is that good sclang code should (and in many cases does) read like documentation.

**Comment by aspiers** [comment]
I agree to the extent that it is possible and indeed very desirable to achieve beautifully legible and often self-explanatory code, mainly through very careful naming of methods and variables, and ensuring that each method does no more than one thing.  However in general that is no substitute for documentation, and hence orthogonal to the issue being discussed here.

**Comment by telephon** [comment]
Yes, you are right of course. It would indeed be good to bring a minimal documentation closer to the code. I'd still prefer it to be in a separate markdown file, however. 

The reason is that if the code reads well, the documentation tends to comes as a disturbance.

**Comment by brianlheim** [comment]
The source file length argument has come up several times before. There is a simple answer, which is comment folding. More editors can do that than can correctly highlight scdoc markup :)

Structuring via a combination of in source and out of source documentation is also very feasible. It's not either/or, both approaches can and do work together. Reiterating one of @aspiers 's original points, I really wonder if our documentation would be as out of sync as it is if it were done in-source. There's also the fact that these docs are unfriendly to edit, but that's a different story. (Btw we have the ability to lint for documentation rendering issues and haven't yet made use of it in CI.)

I think the community at this point generally agrees that scdoc needs to be somehow phased out or seriously altered/augmented. How that will be done is still up for debate.

**Comment by muellmusik** [comment]
>I really wonder if our documentation would be as out of sync as it is if it were done in-source.

Of course it would. The problem is not that people don’t know they need to document stuff or that everyone forgets. It’s a cultural one. Basically no changes should be merged without corresponding doc updates. This has to be the expected thing and devs need to buy into that. 

We’ve made progress on the cultural front but there’s still a lot of old undocumented or out of sync stuff. I think a big push into fixing that would be far more productive in terms of getting stuff up to date than  reinventing the wheel, and I suspect partial reinvention would make that issue worse.

**Comment by brianlheim** [comment]
> I think a big push into fixing that would be far more productive in terms of getting stuff up to date than reinventing the wheel, and I suspect partial reinvention would make that issue worse.

OTOH most of the problems we have right now are because of NIH. We actually don't have to reinvent the wheel. I think I've mentioned this before - writing a custom parser hooking into Doxygen would be much less maintenance effort and immediately give us a lot more flexibility in documentation. Converting scdoc markup to Markdown or out of source Doxygen is not difficult either. There's also a decent ecosystem of tools out there for styling the XML and HTML output.

**Comment by muellmusik** [comment]
Sorry, meant reinventing the wheel in the sense of a big overhaul, not a new bespoke solution.

I think documentation format, inline vs. separate, and current incomplete or outdated state are largely, if not completely separate issues, and I find the way they all get mixed up in these discussions a bit unhelpful. I can see there might be a case for changing scdoc for something else, providing it could be done completely rather than piecemeal, though the question is to what. Personally I don't find doxygen easier than scdoc, and markdown lacks semantic markup I think? (Or does it? with all the flavours it's hard to be sure...)
